packages/slate-locales-plugin/index.js

```diff
@@ -0,0 +1,183 @@
+const fs = require('fs-extra');
+const path = require('path');
+const {ConcatSource, RawSource} = require('webpack-sources');
+const _ = require('lodash');
+
+module.exports = class localesPlugin {
+  constructor(options) {
+    this.options = options;
+  }
+  apply(compiler) {
+    compiler.hooks.compilation.tap('Slate Locales Plugin', this.addLocales);
+  }
+
+  async addLocales(compilation) {
+    const dirents = await fs.readdir(
+      path.resolve(compilation.options.context, 'sections'),
+      {withFileTypes: true},
+    );
+    // eslint-disable-next-line prefer-arrow-callback
+    dirents.forEach((dirent) => {
+      if (dirent.isDirectory()) {
+        traverseDirectory(
+          dirent,
+          path.resolve(compilation.options.context, 'sections'),
+          compilation,
+          dirents,
+        );
+      } else if (dirent.isFile() && path.extname(dirent.name) === '.liquid') {
+        getSchema(
+          dirent,
+          path.resolve(compilation.options.context, 'sections'),
+          compilation,
+          dirents,
+        );
+      }
+    });
+  }
+};
+
+async function getSchema(liquidDirent, directoryPath, compilation, dirents) {
+  if (liquidDirent.name === 'template.liquid') {
+    const outputKey = `../sections${
+      directoryPath.split('/sections')[1]
+    }.liquid`;
+
+    if (
+      dirents.find((obj) => obj.name === 'locales' && obj.isDirectory()) &&
+      dirents.find((obj) => obj.name === 'schema.json' && obj.isFile())
+    ) {
+      compilation.assets[outputKey] = new ConcatSource(
+        new RawSource(
+          fs.readFileSync(path.resolve(directoryPath, 'template.liquid')),
+        ),
+        new RawSource(
+          `{% schema %}\n ${JSON.stringify(
+            createMainSchema(
+              combineLocales(path.resolve(directoryPath, 'locales')),
+              path.resolve(directoryPath, 'schema.json'),
+            ),
+            null,
+            2,
+          )}{% endschema %}`,
+        ),
+      );
+    } else if (
+      dirents.find((obj) => obj.name === 'schema.json' && obj.isFile())
```

Comment:
These conditions are really vague at indicating what transformations you're about to do.